library/psa_crypto.c

```diff
@@ -1165,6 +1216,65 @@ static psa_status_t psa_save_generated_persistent_key( psa_key_slot_t *slot,
 }
 #endif /* defined(MBEDTLS_PSA_CRYPTO_STORAGE_C) */
 
+static psa_status_t psa_copy_key_material( const psa_key_slot_t *source,
+                                           psa_key_handle_t target )
+{
+    psa_status_t status;
+    uint8_t *buffer = NULL;
+    size_t buffer_size = 0;
+    size_t length;
+
+    buffer_size = PSA_KEY_EXPORT_MAX_SIZE( source->type,
+                                           psa_get_key_bits( source ) );
+    buffer = mbedtls_calloc( 1, buffer_size );
+    if( buffer == NULL )
+    {
+        status = PSA_ERROR_INSUFFICIENT_MEMORY;
+        goto exit;
+    }
+    status = psa_internal_export_key( source, buffer, buffer_size, &length, 0 );
+    if( status != PSA_SUCCESS )
+        goto exit;
+    status = psa_import_key( target, source->type, buffer, length );
+
+exit:
+    return( status );
+}
+
+psa_status_t psa_copy_key(psa_key_handle_t source_handle,
+                          psa_key_handle_t target_handle,
+                          const psa_key_policy_t *constraint)
+{
+    psa_key_slot_t *source_slot = NULL;
+    psa_key_slot_t *target_slot = NULL;
+    psa_key_policy_t new_policy;
+    psa_status_t status;
+    status = psa_get_key_from_slot( source_handle, &source_slot, 0, 0 );
+    if( status != PSA_SUCCESS )
+        return( status );
+    status = psa_get_empty_key_slot( target_handle, &target_slot );
+    if( status != PSA_SUCCESS )
+        return( status );
+
+    new_policy = target_slot->policy;
+    status = psa_restrict_key_policy( &new_policy, &source_slot->policy );
+    if( status != PSA_SUCCESS )
+        return( status );
+    if( constraint != NULL )
```

Comment:
`constraint` is mostly redundant with setting the policy on the target handle. If your use case isn't one where you can set the policy on the target handle, it looks useless. That's why it's optional.

I don't see any risk of misuse here because there's already a policy on the target slot. Setting a policy constraint during the copy is only useful when this function is called by a key sharing API where the sharer (who owns the source slot) can't otherwise atomically set the policy on the target slot. The policy argument to the sharing function should not be optional.

I considered making `constraint` override the policy on the target handle, but it seemed more likely to be error-prone. The normal sequence is 1: set policy, 2: create key material. If creating the key material with `psa_copy_key` overrode the policy, that would be surprising, and yet would likely go undetected because the resulting key would be more permissive than intended.